File: packages/cli/src/templates/_shared/lib/context/plan_manager.py
"""Plan lifecycle management — archival, lookup, and path extraction.

Provides pure-data operations on plan files:
- archive_plan: copy plan to context plans/ folder, compute hash + signature
- find_latest_plan: locate the most relevant plan for a context
- extract_plan_path_from_result: parse plan path from ExitPlanMode output

This module does NOT modify mode or state.json.  The calling hook
(e.g. archive_plan.py) is responsible for updating mode via
context_store.update_mode() after archival succeeds.
"""
import hashlib
import json
import re
import uuid
from datetime import datetime
from pathlib import Path
from typing import List, Optional, Tuple

from ..base.atomic_write import atomic_write
from ..base.constants import get_context_dir, get_context_plans_dir
from ..base.logger import log_debug, log_info, log_warn, log_error
from ..base.utils import sanitize_title


# ---------------------------------------------------------------------------
# Plan archival
# ---------------------------------------------------------------------------

def archive_plan(
    plan_path: str,
    context_id: str,
    project_root: Path = None,
) -> Tuple[Optional[str], Optional[str], Optional[str]]:
    """Archive a plan file to the context's plans/ folder.

    Copies the plan content to:
        _output/contexts/{context_id}/plans/{date}-{slug}.md

    Computes a content hash and signature for change detection and
    fallback matching after /clear.

    Does NOT modify state.json or mode — the calling hook handles that
    via context_store.update_mode().

    Args:
        plan_path: Path to the source plan file.
        context_id: Target context identifier.
        project_root: Project root directory (default: from env / cwd).

    Returns:
        (archived_path, plan_hash, plan_signature) on success.
        (None, None, None) on any error.
    """
    plan_file = Path(plan_path)
    if not plan_file.exists():
        log_warn("plan_manager", f"Plan file not found: {plan_path}")
        return None, None, None

    # Read plan content
    try:
        content = plan_file.read_text(encoding="utf-8")
    except Exception as e:
        log_error("plan_manager", f"Failed to read plan: {e}")
        return None, None, None

    # Compute hash and signature
    plan_hash = hashlib.sha256(content.encode("utf-8")).hexdigest()[:12]
    plan_signature = content[:200]

    # Ensure plans directory exists
    plans_dir = get_context_plans_dir(context_id, project_root)
    plans_dir.mkdir(parents=True, exist_ok=True)

    # Generate archive filename: YYYY-MM-DD-<slug>.md
    date_str = datetime.now().strftime("%Y-%m-%d-%H%M")
    slug = sanitize_title(plan_file.stem, max_len=30)
    archive_name = f"{date_str}-{slug}.md"
    archive_path = plans_dir / archive_name

    # Handle filename collisions with counter suffix
    counter = 2
    while archive_path.exists():
        archive_name = f"{date_str}-{slug}-{counter}.md"
        archive_path = plans_dir / archive_name
        counter += 1

    # Write archived plan atomically
    success, error = atomic_write(archive_path, content)
    if not success:
        log_error("plan_manager", f"Failed to write archive: {error}")
        return None, None, None

    log_info("plan_manager", f"Archived plan to: {archive_path}")
    return str(archive_path), plan_hash, plan_signature


# ---------------------------------------------------------------------------
# Plan lookup
# ---------------------------------------------------------------------------

def find_latest_plan(
    context_id: str,
    project_root: Path = None,
) -> Optional[str]:
    """Find the most relevant plan file for a context.

    Priority:
    1. state.json plan_path — if the file still exists on disk.
    2. Most recent .md in plans/ directory by modification time.
    3. None if no plans found.

    Args:
        context_id: Context identifier.
        project_root: Project root directory (default: from env / cwd).

    Returns:
        Absolute path string to the plan file, or None.
    """
    # 1. Check state.json plan_path first
    try:
        from .context_store import load_state
        state = load_state(context_id, project_root)
        if state and state.plan_path:
            plan_path = Path(state.plan_path)
            if plan_path.exists():
                return str(plan_path)
    except Exception as e:
        log_warn("plan_manager", f"Failed to check state.json plan_path: {e}")

    # 2. Fall back to most recent .md in plans/ dir by mtime
    plans_dir = get_context_plans_dir(context_id, project_root)
    if plans_dir.exists():
        plans = sorted(
            plans_dir.glob("*.md"),
            key=lambda p: p.stat().st_mtime,
            reverse=True,
        )
        if plans:
            return str(plans[0])

    # 3. No plan found
    return None


# ---------------------------------------------------------------------------
# Plan identification and normalization
# ---------------------------------------------------------------------------

def generate_plan_id() -> str:
    """Generate a short unique plan identifier (8 hex chars)."""
    return uuid.uuid4().hex[:8]


def normalize_plan_content(text: str) -> str:
    """Aggressively normalize plan content for hashing.

    Strips all XML/HTML tags and collapses whitespace so that
    wrapper variations (e.g. <system-reminder>) don't affect the hash.
    """
    text = re.sub(r'<[^>]+>', '', text)
    text = re.sub(r'\s+', ' ', text).strip()
    return text


def extract_plan_anchors(content: str, max_anchors: int = 5) -> List[str]:
    """Extract structural anchors from plan content.

    Returns markdown headings + first substantial paragraph as short strings.
    Used for fuzzy matching when hash-based matching fails.
    """
    anchors = []
    for line in content.splitlines():
        line = line.strip()
        if line.startswith('#') and len(line) > 3:
            anchors.append(line[:80])
        elif not anchors and len(line) > 20:
            anchors.append(line[:80])
        if len(anchors) >= max_anchors:
            break
    return anchors


# ---------------------------------------------------------------------------
# Transcript-based plan path extraction
# ---------------------------------------------------------------------------

_MAX_TRANSCRIPT_SIZE = 50 * 1024 * 1024  # 50 MB

def find_plan_path_in_transcript(transcript_path: str) -> Optional[str]:
    """Find the plan file path by parsing the session transcript JSONL.

    Searches the transcript in reverse for the most recent Write tool call
    whose file_path targets a .claude/plans/ directory. This is deterministic:
    exact structural match on Path.parts, no fuzzy matching.

    Args:
        transcript_path: Absolute path to the transcript JSONL file.

    Returns:
        The file_path string from the Write tool call, or None.
    """
    if not transcript_path:
        return None

    tp = Path(transcript_path)
    if not tp.exists():
        log_debug("plan_manager", f"Transcript not found: {transcript_path}")
        return None

    try:
        size = tp.stat().st_size
    except OSError:
        return None

    if size > _MAX_TRANSCRIPT_SIZE:
        log_warn("plan_manager", f"Transcript too large ({size} bytes), skipping")
        return None

    try:
        lines = tp.read_text(encoding="utf-8").splitlines()
    except Exception as e:
        log_warn("plan_manager", f"Failed to read transcript: {e}")
        return None

    for line in reversed(lines):
        line = line.strip()
        if not line:
            continue
        try:
            data = json.loads(line)
        except (json.JSONDecodeError, ValueError):
            continue

        content = None
        try:
            content = data["message"]["content"]
        except (KeyError, TypeError):
            continue

        if not isinstance(content, list):
            continue

        for block in content:
            if not isinstance(block, dict):
                continue
            if block.get("type") != "tool_use" or block.get("name") != "Write":
                continue
            file_path = None
            try:
                file_path = block["input"]["file_path"]
            except (KeyError, TypeError):
                continue
            if not file_path:
                continue

            # Check if path contains .claude/plans/ as consecutive parts
            parts = Path(file_path).parts
            for i in range(len(parts) - 1):
                if parts[i] == ".claude" and parts[i + 1] == "plans":
                    log_info("plan_manager", f"Extracted plan path from transcript: {file_path}")
                    return file_path

    log_debug("plan_manager", "No plan Write found in transcript")
    return None


# ---------------------------------------------------------------------------
# Path extraction from tool output
# ---------------------------------------------------------------------------

def extract_plan_path_from_result(tool_result: str) -> Optional[str]:
    """Extract plan file path from ExitPlanMode tool result.

    Parses the pattern: "Your plan has been saved to: <path>"
    from the tool_result string returned by ExitPlanMode.

    Args:
        tool_result: Raw text output from the ExitPlanMode tool.

    Returns:
        Plan file path string (stripped), or None if not found.
    """
    if not tool_result:
        return None
    match = re.search(r"Your plan has been saved to:\s*(.+\.md)", tool_result)
    if match:
        return match.group(1).strip()
    return None

File: packages/cli/src/templates/core/lib-ts/context/plan-manager.ts
/**
 * Plan lifecycle management — archival, lookup, and path extraction.
 * See SPEC.md §9
 *
 * Provides pure-data operations on plan files:
 * - archivePlan: copy plan to context plans/ folder, compute hash + signature
 * - findLatestPlan: locate the most relevant plan for a context
 * - extractPlanPathFromResult: parse plan path from ExitPlanMode output
 */

import * as crypto from "node:crypto";
import * as fs from "node:fs";
import path from "node:path";

import { atomicWrite } from "../runtime/atomic-write.js";
import { getContextPlansDir, sanitizeTitle } from "../runtime/constants.js";
import { logDebug, logInfo, logWarn, logError } from "../runtime/logger.js";
import { generateSlug } from "../runtime/utils.js";

// ---------------------------------------------------------------------------
// Plan archival
// ---------------------------------------------------------------------------

/**
 * Archive a plan file to the context's plans/ folder.
 * Computes a content hash and signature.
 * Does NOT modify state.json or mode.
 * See SPEC.md §9.2
 *
 * Returns [archivedPath, planHash, planSignature] on success,
 * or [null, null, null] on error.
 */
export function archivePlan(
  planPath: string,
  contextId: string,
  projectRoot?: string,
): [string | null, string | null, string | null] {
  if (!fs.existsSync(planPath)) {
    logWarn("plan_manager", `Plan file not found: ${planPath}`);
    return [null, null, null];
  }

  let content: string;
  try {
    content = fs.readFileSync(planPath, "utf8");
  } catch (error_: unknown) {
    logError("plan_manager", `Failed to read plan: ${error_}`);
    return [null, null, null];
  }

  // Compute hash and signature
  const planHash = crypto.createHash("sha256").update(content, "utf8").digest("hex").slice(0, 12);
  const planSignature = content.slice(0, 200);

  // Ensure plans directory exists
  const plansDir = getContextPlansDir(contextId, projectRoot);
  fs.mkdirSync(plansDir, { recursive: true });

  // Generate archive filename: YYYY-MM-DD-HHMM-<slug>.md
  const now = new Date();
  const dateStr = [
    now.getFullYear(),
    "-",
    String(now.getMonth() + 1).padStart(2, "0"),
    "-",
    String(now.getDate()).padStart(2, "0"),
    "-",
    String(now.getHours()).padStart(2, "0"),
    String(now.getMinutes()).padStart(2, "0"),
  ].join("");

  // Extract a clean summary from plan content for slug generation.
  // Headings describe the plan's intent better than raw markdown body.
  const summary = extractPlanSummary(content);
  const slug = generateSlug(summary, 60, sanitizeTitle(
    path.basename(planPath, path.extname(planPath)), 30,
  ));

  let archiveName = `${dateStr}-${slug}.md`;
  let archivePath = path.join(plansDir, archiveName);

  // Handle filename collisions
  let counter = 2;
  while (fs.existsSync(archivePath)) {
    archiveName = `${dateStr}-${slug}-${counter}.md`;
    archivePath = path.join(plansDir, archiveName);
    counter++;
  }

  // Write archived plan atomically
  const [success, error] = atomicWrite(archivePath, content);
  if (!success) {
    logError("plan_manager", `Failed to write archive: ${error}`);
    return [null, null, null];
  }

  logInfo("plan_manager", `Archived plan to: ${archivePath}`);
  return [archivePath, planHash, planSignature];
}

/**
 * Extract a human-readable summary from plan markdown content.
 * Pulls headings and the first substantial paragraph, producing
 * text suitable for the AI slug generator (which expects conversational input).
 */
function extractPlanSummary(content: string): string {
  const lines = content.split(/\r?\n/);
  const parts: string[] = [];
  let firstParagraph = "";

  for (const line of lines) {
    const trimmed = line.trim();
    // Collect markdown headings (strip # prefix)
    if (trimmed.startsWith("#")) {
      const heading = trimmed.replace(/^#+\s*/, "");
      if (heading.length > 2) parts.push(heading);
    }
    // Grab first substantial non-heading line as context
    if (!firstParagraph && !trimmed.startsWith("#") && trimmed.length > 20) {
      firstParagraph = trimmed.slice(0, 120);
    }
    // Enough material for the AI
    if (parts.length >= 5) break;
  }

  if (firstParagraph) parts.push(firstParagraph);
  return parts.join(" ").slice(0, 500) || content.slice(0, 500);
}

// ---------------------------------------------------------------------------
// Plan lookup
// ---------------------------------------------------------------------------

/**
 * Find the most relevant plan file for a context.
 * Priority: state.json plan_path > most recent .md in plans/ dir.
 * See SPEC.md §9.3
 */
export function findLatestPlan(
  contextId: string,
  projectRoot?: string,
): string | null {
  // 1. Check state.json plan_path first
  try {
    // Dynamic import to avoid circular dependency at module level
    // eslint-disable-next-line @typescript-eslint/no-require-imports, no-undef -- dynamic require to avoid circular dependency
    const stateIo = require("../runtime/state-io.js");
    const state = stateIo.readStateJson(contextId, projectRoot);
    if (state?.plan_path && fs.existsSync(state.plan_path)) {
      return state.plan_path;
    }
  } catch (error: unknown) {
    logWarn("plan_manager", `Failed to check state.json plan_path: ${error}`);
  }

  // 2. Fall back to most recent .md in plans/ dir
  const plansDir = getContextPlansDir(contextId, projectRoot);
  if (fs.existsSync(plansDir)) {
    try {
      const files = fs.readdirSync(plansDir)
        .filter(f => f.endsWith(".md"))
        .map(f => {
          const fullPath = path.join(plansDir, f);
          return { path: fullPath, mtime: fs.statSync(fullPath).mtimeMs };
        })
        .sort((a, b) => b.mtime - a.mtime);

      if (files.length > 0) {
        return files[0]!.path;
      }
    } catch { /* ignore */ }
  }

  return null;
}

// ---------------------------------------------------------------------------
// Plan identification and normalization
// ---------------------------------------------------------------------------

/**
 * Generate a short unique plan identifier (8 hex chars).
 * See SPEC.md §9.4
 */
export function generatePlanId(): string {
  return crypto.randomUUID().replaceAll('-', "").slice(0, 8);
}

/**
 * Aggressively normalize plan content for hashing.
 * Strips all XML/HTML tags and collapses whitespace.
 * See SPEC.md §9.5
 */
export function normalizePlanContent(text: string): string {
  let result = text.replaceAll(/<[^>]+>/g, "");
  result = result.replaceAll(/\s+/g, " ").trim();
  return result;
}

/**
 * Extract structural anchors from plan content.
 * Returns markdown headings + first substantial paragraph as short strings.
 * See SPEC.md §9.6
 */
export function extractPlanAnchors(content: string, maxAnchors = 5): string[] {
  const anchors: string[] = [];
  for (const line of content.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (trimmed.startsWith("#") && trimmed.length > 3) {
      anchors.push(trimmed.slice(0, 80));
    } else if (anchors.length === 0 && trimmed.length > 20) {
      anchors.push(trimmed.slice(0, 80));
    }
    if (anchors.length >= maxAnchors) break;
  }
  return anchors;
}

// ---------------------------------------------------------------------------
// Transcript-based plan path extraction
// ---------------------------------------------------------------------------

const MAX_TRANSCRIPT_SIZE = 50 * 1024 * 1024; // 50 MB

/**
 * Find the plan file path by parsing the session transcript JSONL.
 * Searches in reverse for the most recent Write tool call targeting .claude/plans/.
 * See SPEC.md §9.7
 */
export function findPlanPathInTranscript(transcriptPath: string): string | null {
  if (!transcriptPath) return null;

  if (!fs.existsSync(transcriptPath)) {
    logDebug("plan_manager", `Transcript not found: ${transcriptPath}`);
    return null;
  }

  let size: number;
  try {
    size = fs.statSync(transcriptPath).size;
  } catch {
    return null;
  }

  if (size > MAX_TRANSCRIPT_SIZE) {
    logWarn("plan_manager", `Transcript too large (${size} bytes), skipping`);
    return null;
  }

  let lines: string[];
  try {
    lines = fs.readFileSync(transcriptPath, "utf8").split(/\r?\n/);
  } catch (error: unknown) {
    logWarn("plan_manager", `Failed to read transcript: ${error}`);
    return null;
  }

  for (let i = lines.length - 1; i >= 0; i--) {
    const line = lines[i]!.trim();
    if (!line) continue;

    let data: unknown;
    try {
      data = JSON.parse(line);
    } catch {
      continue;
    }

    let contentArr: unknown;
    try {
      const message = (data as { message?: { content?: unknown } }).message;
      contentArr = message?.content;
    } catch {
      continue;
    }

    if (!Array.isArray(contentArr)) continue;

    for (const block of contentArr) {
      if (typeof block !== "object" || block === null) continue;
      if (block.type !== "tool_use" || block.name !== "Write") continue;

      const filePath = block.input?.file_path;
      if (!filePath) continue;

      // Check if path contains .claude/plans/ as consecutive parts
      const parts = filePath.replaceAll('\\', "/").split("/");
      for (let j = 0; j < parts.length - 1; j++) {
        if (parts[j] === ".claude" && parts[j + 1] === "plans") {
          logInfo("plan_manager", `Extracted plan path from transcript: ${filePath}`);
          return filePath;
        }
      }
    }
  }

  logDebug("plan_manager", "No plan Write found in transcript");
  return null;
}

// ---------------------------------------------------------------------------
// Path extraction from tool output
// ---------------------------------------------------------------------------

/**
 * Extract plan file path from ExitPlanMode tool result.
 * Parses the pattern: "Your plan has been saved to: <path>"
 * See SPEC.md §9.8
 */
export function extractPlanPathFromResult(toolResult: string): string | null {
  if (!toolResult) return null;
  const match = toolResult.match(/Your plan has been saved to:\s*(.+\.md)/);
  return match ? match[1]!.trim() : null;
}



